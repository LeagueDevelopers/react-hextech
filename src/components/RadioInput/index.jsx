import React, { PropTypes, PureComponent } from 'react';
import cx from 'classnames';

import withOptions from '../../util/withOptions';

import Option from '../RadioOption';

import style from './index.scss';

let instanceId = 0;

/**
 * RadioInput Component
 *
 * @export
 * @class RadioInput
 * @extends {PureComponent}
 */
@withOptions
export default class RadioInput extends PureComponent {
  static propTypes = {
    className: PropTypes.any,
    disabled: PropTypes.bool,
    value: PropTypes.any,
    label: PropTypes.string,
    options: PropTypes.array,
    onChange: PropTypes.func,
    onBlur: PropTypes.func
  }

  static defaultProps = {
    className: undefined,
    disabled: false,
    value: undefined,
    label: '',
    options: [],
    onChange: Function.prototype,
    onBlur: Function.prototype
  }

  constructor (props) {
    super(props);

    instanceId += 1;
    this.instanceId = instanceId;
  }

  handleSelect = (nextValue, evt) => {
    evt.preventDefault();
    const { value, onChange } = this.props;
    if (nextValue !== value) {
      onChange(nextValue);
    }
  }

  handleBlur = evt => {
    const { disabled, onBlur } = this.props;
    if (!disabled) {
      onBlur(evt);
    }
  }

  render () {
    const { options, className, disabled, value, label } = this.props;

    return (
      <div role="radiogroup" aria-labelledby={label} className={cx(style.radioInput, className)}>
        {options.map((o, i) =>
          <Option
            key={`${this.instanceId}${o.label}`}
            checked={o.value === value}
            disabled={disabled || o.disabled}
            onChange={evt => this.handleSelect(o.value, evt)}
            onBlur={this.handleBlur}
            label={o.label}
            value={o.value}
            tabIndex={i > 0 ? -1 : 0}
          />)}
      </div>
    );
  }
}
