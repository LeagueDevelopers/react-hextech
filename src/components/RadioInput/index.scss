.radioInput {
	display: flex;
	flex-direction: column;
	justify-content: center;
	outline: none;
}
