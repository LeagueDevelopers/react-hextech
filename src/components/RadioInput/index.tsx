import React, { PureComponent } from 'react';
import cx from 'classnames';

import { withOptions } from '@utils';

import RadioOption from '../RadioOption';
import { Option } from '../option';

const style = require('./index.scss');

interface RadioInputProps<T = any> {
  className?: any;
  disabled?: boolean;
  value: any;
  label?: string;
  options: Option<T>[];
  onChange: (nextValue: any) => void;
  onBlur?: () => void;
}

class RadioInput extends PureComponent<RadioInputProps> {
  static defaultProps = {
    className: undefined,
    disabled: false,
    value: undefined,
    label: '',
    options: [],
    onChange: Function.prototype,
    onBlur: Function.prototype,
  };

  handleSelect = (nextValue, evt) => {
    evt.preventDefault();
    const { value, onChange } = this.props;

    if (nextValue !== value) {
      onChange(nextValue);
    }
  };

  render() {
    const { options, className, disabled, value, label, onBlur } = this.props;

    return (
      <div
        role="radiogroup"
        aria-labelledby={label}
        className={cx(style.radioInput, className)}
      >
        {options.map((o, i) => (
          <RadioOption
            key={`${o.label}`}
            name={`group-option`}
            checked={o.value === value}
            disabled={disabled || o.disabled}
            onChange={evt => this.handleSelect(o.value, evt)}
            onBlur={onBlur}
            label={o.label}
            value={o.value}
            tabIndex={i > 0 ? -1 : 0}
          />
        ))}
      </div>
    );
  }
}

export default withOptions(RadioInput);
