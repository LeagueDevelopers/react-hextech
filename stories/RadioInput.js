import React from 'react';
import { storiesOf, action } from '@storybook/react';
import { RadioInput as BaseRadioInput, stateful } from '../src';

const RadioInput = stateful(BaseRadioInput);

const options = ['Jinx', 'Leona', 'Renekton', 'Quinn'];

storiesOf('Radio Input', module)
  .add('with options', () => (
    <RadioInput onChange={action('radio-input-change')} onBlur={action('radio-input-blur')} options={options} />
  ))
  .add('disabled', () => (
    <RadioInput disabled onChange={action('radio-input-change')} onBlur={action('radio-input-blur')} options={options} />
  ));
