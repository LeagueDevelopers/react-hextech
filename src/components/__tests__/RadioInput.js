import React from 'react';
import { shallow, mount } from 'enzyme';

import RadioInput from '../RadioInput';

const options = [
  { value: 0, label: 'test0' },
  { value: 1, label: 'test1' },
  { value: 2, label: 'test2' },
  { value: 5, label: 'test3' },
];

describe.only('<RadioInput />', () => {
  it('should render correctly', () => {
    const props = {
      onChange: jest.fn(),
      onBlur: jest.fn(),
      selectedOption: options[1],
      value: options[1].value,
    };
    const wrapper = shallow(<RadioInput options={options} {...props}>Hello</RadioInput>);
    expect(wrapper.get(0)).toMatchSnapshot();
  });

  it('should render as many children as the options it is provided', () => {
    const wrapper = mount(<RadioInput options={options} />);
    expect(wrapper.children().length).toBe(options.length);
  });

  it('should render no children if no options are provided', () => {
    const wrapper = shallow(<RadioInput options={[]} />);
    expect(wrapper.children().length).toBe(0);
  });

  it('should render no children if options is undefined or otherwise lacks a map method', () => {
    const wrapper = shallow(<RadioInput />);
    expect(wrapper.children().length).toBe(0);
  });

  it('should render the correct label for each child', () => {
    const wrapper = shallow(<RadioInput options={options} />);
    wrapper.children().forEach((child, i) => {
      expect(child.props().label).toBe(options[i].label);
    });
  });

  it('should pass the disabled prop to its options', () => {
    const wrapper = shallow(<RadioInput options={options} disabled />);
    wrapper.children().forEach(child => {
      expect(child.props().disabled).toBe(true);
    });
  });

  it('should mark the selected option, using strict equality against the value prop', () => {
    const selectedOption = options[1];
    const value = selectedOption.value;

    const wrapper = mount(<RadioInput options={options} value={value} />);
    const selected = wrapper.childAt(1);

    const unselectedNodes = wrapper.children()
      .map(child => child.props())
      .filter(child => !child.checked);

    expect(selected.props().checked).toBe(true);
    expect(unselectedNodes).toHaveLength(options.length - 1);
  });

  it('should call onChange with the new value if the selection changes', () => {
    const handleChange = jest.fn();
    const selectedOption = options[1];
    const value = selectedOption.value;

    const wrapper = mount(<RadioInput options={options} value={value} onChange={handleChange} />);

    handleChange.mockImplementation(v => wrapper.setProps({ value: v }));

    const nextSelect = wrapper.childAt(3).find('input');

    nextSelect.simulate('change');
    expect(handleChange).toHaveBeenCalledWith(options[3].value);

    nextSelect.simulate('change');
    expect(handleChange).toHaveBeenCalledTimes(1);
  });

  it('should not call onChange with the new value if the selection does not change', () => {
    const handleChange = jest.fn();
    const selectedOption = options[1];
    const value = selectedOption.value;

    const wrapper = mount(<RadioInput options={options} value={value} onChange={handleChange} />);

    handleChange.mockImplementation(v => wrapper.setProps({ value: v }));

    const nextSelect = wrapper.childAt(1).find('input');

    nextSelect.simulate('change');
    expect(handleChange).not.toHaveBeenCalled();
  });

  it('should fail gracefully if neither onClick or onChange are defined', () => {
    const selectedOption = options[1];
    const value = selectedOption.value;

    const wrapper = shallow(<RadioInput options={options} value={value} />);

    const nextSelect = wrapper.childAt(3);

    nextSelect.simulate('click');
  });
});
