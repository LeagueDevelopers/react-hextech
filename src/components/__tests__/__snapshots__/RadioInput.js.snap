// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioInput /> should render correctly 1`] = `
<RadioInput
  className={undefined}
  disabled={false}
  label=""
  onBlur={[Function]}
  onChange={[Function]}
  options={
    Array [
      Object {
        "hextech__label": "test0",
        "label": "test0",
        "value": Object {
          "label": "test0",
          "value": 0,
        },
      },
      Object {
        "hextech__label": "test1",
        "label": "test1",
        "value": 1,
      },
      Object {
        "hextech__label": "test2",
        "label": "test2",
        "value": 2,
      },
      Object {
        "hextech__label": "test3",
        "label": "test3",
        "value": 5,
      },
    ]
  }
  selectedOption={
    Object {
      "label": "test1",
      "value": 1,
    }
  }
  value={1}
>
  Hello
</RadioInput>
`;
